lib/api/apiUtils/object/versioning.js

```diff
@@ -144,29 +185,48 @@ export function versioningPreprocessing(bucketName, bucketMD, objectKey,
             // object does not exist or is not versioned (before versioning)
             options.versionId = '';
             options.isNull = true;
-            options.deleteData = true;
+            options.dataToDelete = mstObjLocation;
             return callback(null, options);
         }
         // TODO check AWS behaviour
         return callback(errors.BadRequest);
+    // NOTE: Currently we do not accept receive requests with PUTs to
+    // specific version ID, but this may change after implementing
+    // georeplication.
     } else if (mstIsNull) {
         // master is versioned and is a null version
         if (reqVersionId === 'null') {
             // overwrite the existing version, make new version null
             options.versionId = '';
             options.isNull = true;
-            options.deleteData = true;
+            options.dataToDelete = mstObjLocation;
             return callback(null, options);
         }
-        // TODO check AWS behaviour
         options.versionId = reqVersionId;
-        options.deleteData = true;
-        return callback(null, options);
+        return _getVersionLocation(bucketName, objectKey,
+            { versionId: reqVersionId }, log, (err, dataToDelete) => {
+                if (err && err === 'NoSuchVersion') {
+                    // no data to delete
+                    return callback(null, options);
+                } else if (err) {
+                    return callback(errors.InternalError);
```

Comment:
Mapped because the only error we expect is 'NoSuchKey'; if it is some other system error, we won't expose it to the user. I added another log.